File: TwoPointers/Java/1_ValidPalindrome.java
/**
https://leetcode.com/problems/valid-palindrome/

A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.

Given a string s, return true if it is a palindrome, or false otherwise.
*/

class Solution {
    public boolean isPalindrome(String s) {
        int left = 0, right = s.length() - 1;
        while (left < right) {
            if (!Character.isLetterOrDigit(s.charAt(left))) left++;
            else if (!Character.isLetterOrDigit(s.charAt(right))) right--;
            else {
                if (Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right))) return false;
                left++;
                right--;
            }
        }
        return true;
    }
}

/**
Algorithm: Use two pointers to traverse the left and right directions simultaneously. If the char is not alpha-numeric then go to the next index in that direction. Otherwise convert the char to lower then do a comparison. Stop when the left pointer value exceeds the right pointer value

Space: O(1), no extra space used
Time: O(n), one pass through the string, two pointers both going towards the center
*/

File: TwoPointers/C#/1_ValidPalindrome.cs
/**
https://leetcode.com/problems/valid-palindrome/

A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.

Given a string s, return true if it is a palindrome, or false otherwise.
*/

public class Solution {
    public bool IsPalindrome(string s) {
        int start = 0, end = s.Length - 1;

        while (start < end) {
            if (!Char.IsLetterOrDigit(s[start])) start++;
            else if (!Char.IsLetterOrDigit(s[end])) end--;
            else {
                if (Char.ToLower(s[start]) != Char.ToLower(s[end])) return false;
                start++;
                end--;
            }
        }
        return true;
    }
}

/**
Algorithm:
Use two pointers, one at the end and the other at the beginning. Compare the lower case chars one by one only if they're a letter or digit. Otherwise skip that iteration

Space: O(1), no extra space used
Time: O(n), all chars of the string compared
*/